src/uct/ib/rdmacm/rdmacm_cm_ep.c

```diff
@@ -240,31 +262,63 @@ static ucs_status_t
 uct_rdmacm_cm_ep_create_reserved_qpn(uct_rdmacm_cm_ep_t *cep,
                                      uct_rdmacm_cm_device_context_t *ctx)
 {
-    uint32_t qpns_per_obj = UCS_BIT(ctx->log_reserved_qpn_granularity);
+    uint32_t qpns_per_obj                      =
+        UCS_BIT(ctx->log_reserved_qpn_granularity);
+    uct_rdmacm_cm_peer_dev_ctx_t *peer_dev_ctx = NULL;
     uct_rdmacm_cm_reserved_qpn_blk_t *blk;
     ucs_status_t status;
 
     ucs_spin_lock(&ctx->lock);
 
-    blk = ucs_list_tail(&ctx->blk_list, uct_rdmacm_cm_reserved_qpn_blk_t, entry);
-    if (ucs_list_is_empty(&ctx->blk_list) ||
-        (blk->next_avail_qpn_offset >= qpns_per_obj)) {
-        status = uct_rdmacm_cm_reserved_qpn_blk_alloc(ctx, cep->id->verbs,
-                                                      UCS_LOG_LEVEL_ERROR,
-                                                      &blk);
+    if (ucs_unlikely(ucs_list_is_empty(&ctx->blk_list))) {
+        RDMACM_CM_RESERVE_QPN_BLK(ctx, cep, status, blk, { goto out; });
+    }
+
+    if (ctx->reuse_qpn) {
+        status = uct_rdmacm_cm_get_peer_dev_ctx(ctx, &cep->id->route,
+                                                &peer_dev_ctx);
         if (status != UCS_OK) {
             goto out;
         }
 
-        ucs_list_add_tail(&ctx->blk_list, &blk->entry);
+        if (peer_dev_ctx->ref_qpn_blk == NULL) {
+            blk                                 = ucs_list_next(&ctx->blk_list,
+                                                                uct_rdmacm_cm_reserved_qpn_blk_t, entry);
+            peer_dev_ctx->ref_qpn_blk           = blk;
+            peer_dev_ctx->next_avail_qpn_offset = 0;
+        } else if (peer_dev_ctx->next_avail_qpn_offset >= qpns_per_obj) {
+            if (peer_dev_ctx->ref_qpn_blk->entry.next == &ctx->blk_list) {
+                RDMACM_CM_RESERVE_QPN_BLK(ctx, cep, status, blk, { goto out; });
+            }
+            blk                                 = ucs_list_next(&peer_dev_ctx->ref_qpn_blk->entry,
+                                                                uct_rdmacm_cm_reserved_qpn_blk_t, entry);
+            peer_dev_ctx->ref_qpn_blk           = blk;
+            peer_dev_ctx->next_avail_qpn_offset = 0;
+        } else {
+            blk = peer_dev_ctx->ref_qpn_blk;
+        }
+    } else {
+        blk = ucs_list_tail(&ctx->blk_list, uct_rdmacm_cm_reserved_qpn_blk_t,
+                            entry);
+        if (blk->next_avail_qpn_offset >= qpns_per_obj) {
+            RDMACM_CM_RESERVE_QPN_BLK(ctx, cep, status, blk, { goto out; });
+        }
     }
 
-    cep->qpn = blk->first_qpn + blk->next_avail_qpn_offset;
+    if (ctx->reuse_qpn) {
+        ucs_assert(peer_dev_ctx != NULL);
+        cep->qpn                   = blk->first_qpn +
+                                     peer_dev_ctx->next_avail_qpn_offset++;
```

Comment:
@evgeny-leksikov Do you want to add comment for this line to explain that ```peer_dev_ctx->next_avail_qpn_offset++``` is for unique peer dev to allocate qpn?